### PR TITLE
Bls integration

### LINKED ISSUE #202: Finish integrating BLS into CDT main branch
https://github.com/AntelopeIO/cdt/pull/154 was merged into the `bls_integration` branch as part of completing the BLS milestone of IF. However, some changes still remain before the code can be merged into main.

First, the branch should be updated to use the leap-dev package generated from the `bls_integration` branch in Leap so that the integration tests can correctly run. Use this to ensure all the tests are running as intended.

Then, when https://github.com/AntelopeIO/leap/issues/1413 is completed, which means BLS support is integrated into the `main` branch of Leap, the `bls_integration` branch should again be updated to use the leap-dev package from the `main` branch of Leap. The code changes of that branch (with CICD all passing) should then be merged into the `main` branch of CDT.
```[tasklist]
### Tasks
- [ ] https://github.com/AntelopeIO/leap/issues/1462
```

### PR DESCRIPTION
**Pull request by dimas1185**
resolves #202

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+            bls_g1_add(lhs, rhs, r);
+            check(0 == memcmp(r, res.data(), sizeof(bls_g1)), "bls_g1_add test failed");
+        }
+
+        [[eosio::action]]
+        void testg2add(const std::vector<uint8_t>& op1, const std::vector<uint8_t>& op2, const std::vector<uint8_t>& res)
+        {
+            bls_g2 lhs, rhs, r;
+            memcpy(lhs, op1.data(), sizeof(bls_g2));
+            memcpy(rhs, op1.data(), sizeof(bls_g2));
```
Comment (greg7mdp) [inline tests/unit/test_contracts/bls_primitives_tests.cpp:220]: Same here should be `op2.data()`.

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+class [[eosio::contract]] bls_primitives_tests : public contract{
+    public:
+        using contract::contract;
+
+        [[eosio::action]]
+        void testg1add(const std::vector<uint8_t>& op1, const std::vector<uint8_t>& op2, const std::vector<uint8_t>& res)
+        {
+            bls_g1 lhs, rhs, r;
+            memcpy(lhs, op1.data(), sizeof(bls_g1));
+            memcpy(rhs, op1.data(), sizeof(bls_g1));
```
[the remark this answers is not in the document]
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:210]: @mschoenebeck can you chime in, please?

[the remark this answers is not in the document]
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:210]: Correct, that is a typo. Should be `op2.data()`...

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+            bls_g2_add(lhs, rhs, r);
+            check(0 == memcmp(r, res.data(), sizeof(bls_g2)), "bls_g2_add test failed");
+        }
+
+        [[eosio::action]]
+        void testg1mul(const std::vector<uint8_t>& point, const std::vector<uint8_t>& scalar, const std::vector<uint8_t>& res)
+        {
+            bls_g1 p, r;
+            bls_scalar s;
+            memcpy(p, point.data(), sizeof(bls_g1));
```
[the remark this answers is not in the document]
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:230]: Yeah, I believe I had trouble with pointer alias errors or something.. I can't really remember. Of course, if `reinterpret_cast` works here, than that would be more efficient.

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:210]: Same in line `220`

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 1 above)
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:220]: yes, same here.

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp]: @greg7mdp @heifner strict aliasing permits downcasting to `char*`, `unsigned char*`. Anyway I refactored the code, removed unnecessary `memcpy` and removed `reinterpret_cast` from `cdt` wrappers

[the remark this answers is not in the document]
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp]: ooops, it seems that I was wrong, something wrong with ubuntu 20 build, let me check

[the remark this answers is not in the document]
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp]: fixed

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+class [[eosio::contract]] bls_primitives_tests : public contract{
+    public:
+        using contract::contract;
+
+        [[eosio::action]]
+        void testg1add(const std::vector<char>& op1, const std::vector<char>& op2, const std::vector<char>& res) {
+            check(op1.size() == std::tuple_size<bls_g1>::value, "wrong op1 size passed");
+            check(op2.size() == std::tuple_size<bls_g1>::value, "wrong op2 size passed");
+            bls_g1 r;
+            bls_g1_add(reinterpret_cast<const bls_g1&>(op1[0]), reinterpret_cast<const bls_g1&>(op2[0]), r);
```
[the remark this answers is not in the document]
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:189]: Huh, is that code from me? If so, I totally agree with your change. Can't believe I wrote that.. seems very wrong to me as well!

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 5]
```diff
@@ -203,118 +182,89 @@ class [[eosio::contract]] bls_primitives_tests : public contract{
...
-            bls_g2_map(element, r);
-            check(0 == memcmp(r, res.data(), sizeof(bls_g2)), "bls_g2_map test failed");
+            bls_g2_map(reinterpret_cast<const bls_fp2&>(e[0]), r);
+            check(std::equal(res.begin(), res.end(), r.begin()), "bls_g2_map test failed");
         }
 
         // Construct an extensible-output function based on SHA256
         void xmd_sh256(
-            uint8_t *buf,
+            char *buf,
```
Comment (greg7mdp) [inline tests/unit/test_contracts/bls_primitives_tests.cpp:257]: I'm not sure why the change from `uint8_t` to `char`, but if we need to change I'd prefer `std::byte` unless there is a reason `char` is needed.

File: tests/unit/test_contracts/bls_primitives_tests.cpp [hunk 6]
```diff
@@ -203,118 +182,89 @@ class [[eosio::contract]] bls_primitives_tests : public contract{
...
             int buf_len,
             const uint8_t *in,
             int in_len,
-            const uint8_t *dst,
+            const char *dst,
```
Comment (greg7mdp) [inline tests/unit/test_contracts/bls_primitives_tests.cpp:261]: same here. Use `std::byte`?

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:189]: other than personal preference I do not know any issues with that but I'll change that if that looks too unusual

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 5 above)
Comment (dimas1185) [reply tests/unit/test_contracts/bls_primitives_tests.cpp:257]: the thing is, we have char inside low level interfaces. so we have to cast it on some level. I'm trying to reduce that.
https://github.com/AntelopeIO/cdt/blob/bls_integration/libraries/eosiolib/core/eosio/crypto_bls_ext.hpp

File: tests/unit/test_contracts/bls_primitives_tests.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mschoenebeck) [reply tests/unit/test_contracts/bls_primitives_tests.cpp]: Sorry if that was a change from you. I didn't mean to be rude or something, just thought I committed this somehow by accident because it looks indeed very unusual to me. No offense!

**Review by greg7mdp — APPROVED** [review]

File: libraries/eosiolib/core/eosio/crypto_bls_ext.hpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            int32_t bls_fp_mod(const char* s, uint32_t s_len, char* res, uint32_t res_len);
+        }
+    }
+
+    using bls_scalar = std::array<char, 32>;
+    using bls_fp     = std::array<char, 48>;
+    using bls_s      = std::array<char, 64>;
+    using bls_fp2    = std::array<bls_fp, 2>;
+    using bls_g1     = std::array<char, 144>;
+    using bls_g2     = std::array<char, 288>;
```
Comment (heifner) [inline libraries/eosiolib/core/eosio/crypto_bls_ext.hpp:51]: Why represent `bls_fp2` as `std::array<bls_fp, 2>`, but not `bls_g2` as `std::array<bls_f2, 3>` ?

File: libraries/eosiolib/core/eosio/crypto_bls_ext.hpp [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    using bls_gt     = std::array<char, 576>;
```
Comment (heifner) [inline libraries/eosiolib/core/eosio/crypto_bls_ext.hpp:52]: Why not `std::array<bls_g2, 2>` ?

**Review by heifner — APPROVED** [review]

File: libraries/eosiolib/core/eosio/crypto_bls_ext.hpp (hunk 7 above)
Comment (dimas1185) [reply libraries/eosiolib/core/eosio/crypto_bls_ext.hpp:51]: @mschoenebeck is better to comment but as per the code `bls_g1` is not used partially same way `bls_fp2` does.
If that will make any sense we will change this as a separate PR.

**Comment by ericpassmore** [comment]
Note:start
group: FEATURE
category: INTERNALS
summary: Add BLS cryptography functions.
Note:end
